main.go

```diff
@@ -165,10 +214,200 @@ func convertToV2(v1 models.CircleYamlV1) (models.CircleYamlV2, error) {
+			"command": `cd $HOME && git clone --depth 1 -v https://github.com/Clever/ci-scripts.git && cd ci-scripts && git show --oneline -s`,
+		},
+	}
+	v2.Jobs.Build.Steps = append(v2.Jobs.Build.Steps, cloneCIScriptsStep)
+}
+
+func determineWorkingDirectory(appType string) (string, error) {
+	// @TODO: determine decent working directory depending on app type for non-(go, wag, node) apps
+	// go, wag: /go/src/github.com/Clever/catapult
+	// non-wag node: ~/Clever/hubble
+
+	org := "Clever"
+
+	// get repo
+	dir, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+
+	splitDir := strings.Split(dir, "/")
+	if len(splitDir) < 1 {
+		return "", fmt.Errorf("failed to find repo in %s", dir)
+	}
+	repo := splitDir[len(splitDir)-1]
+
+	// put together working directory string
+	if appType == GOLANG_APP_TYPE || appType == WAG_APP_TYPE {
+		return fmt.Sprintf("/go/src/github.com/%s/%s", org, repo), nil
+	}
+	return fmt.Sprintf("~/%s/%s", org, repo), nil
+}
+
+func determineImageConstraints() models.ImageConstraints {
+	// if node, will have package.json and node.mk (but this is clever-specific) in main project dir
+	// if go, will have golang.mk (but this is clever-specific)
+	// another common occurance is go with node, which for us is mostly wag
+	imageConstraints := models.ImageConstraints{
+		AppType: "unknown",
+	}
+	if _, err := os.Stat("./swagger.yml"); err == nil {
+		imageConstraints = models.ImageConstraints{
+			AppType: WAG_APP_TYPE,
+			Version: determineGoVersion(),
+		}
+	} else if _, err := os.Stat("./golang.mk"); err == nil {
+		imageConstraints = models.ImageConstraints{
+			AppType: GOLANG_APP_TYPE,
+			Version: determineGoVersion(),
+		}
+	} else if _, err := os.Stat("./node.mk"); err == nil {
+		imageConstraints = models.ImageConstraints{
+			AppType: NODE_APP_TYPE,
+			Version: determineNodeVersion(),
+		}
+	}
+	return imageConstraints
+}
+
+// determineGoVersion determines version of go in use for an app
+// this information is in makefile's golang-version-check, e.g.:
+// $(eval $(call golang-version-check,1.10))
+// @TODO: error if version not found instead of returning 1.10? or is 1.10 default alright?
+func determineGoVersion() string {
+	version := "1.10"
+	versionCheckRegexp, err := regexp.Compile(`golang-version-check,([0-1].[0-9]+)`)
+	if err != nil {
+		fmt.Printf("!WARNING: error compiling regexp %s\n", err.Error())
+	}
+	makefile, err := ioutil.ReadFile("Makefile")
+	if err != nil {
+		log.Fatal(err)
+	}
+	versionCheck := versionCheckRegexp.FindSubmatch(makefile)
+	if versionCheck != nil {
+		version = string(versionCheck[1])
+	}
+	return version
+}
+
+// determineNodeVersion determines version of node for an app
+// @TODO (INFRA-3156): implement (determine correct node version for non-wag node apps)
+func determineNodeVersion() string {
+	version := "8"
+	versionCheckRegexp, err := regexp.Compile(`NODE_VERSION := "v([0-9]+)"`)
+	if err != nil {
+		fmt.Printf("!WARNING: error compiling regexp %s\n", err.Error())
```

Comment:
Since this regex is static, why not just put it in the global context with `regexp.MustCompile`?

Reply:
Switched to `MustCompile` (because I 100% know what the regexp is; why not make sure it actually compiles) but leaving it local for now to keep code related to finding the version in the function that finds the version.  I'm currently not too concerned about efficiency of compiling regexp in the function because function is only evoked once.

Reply:
Updated in 4b65a4cd5c17d25c344e912b9de8cf87217a1e7a